### DISCUSSION TITLE
Button toggle mode?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [alternetsoft/AlternetUI#137] [Ideas]
Is there a way to get grouped toggle buttons?
![image](https://github.com/alternetsoft/AlternetUI/assets/2453707/e24d7f80-3eab-4fc8-b025-73d318b28070)

![image](https://github.com/alternetsoft/AlternetUI/assets/2453707/3ad9dd15-53b1-4e14-9b13-d64d1f1db778)

**Comment by generalloki** [comment]
If you need Button controls, I beleive you can handle Click event and change IsBold for the "active" button in order to show
user what button is "selected".

**Comment by Shadowblitz16** [comment]
ok that will do for a temporary fix

**Comment by generalloki** [comment]
Is "grouped toggle buttons" feature present in Maui or Winforms?

**Comment by Shadowblitz16** [comment]
> Is "grouped toggle buttons" feature present in Maui or Winforms?

I don't know, but I know vb6 win32 forms supported it.

**Comment by generalloki** [comment]
I am closing this for now as button groups can be implemented using Click events.

**Comment by Shadowblitz16** [comment]
@generalloki I looked into this and basicly what I need is a radiobox with the style of a button
https://forums.wxwidgets.org/viewtopic.php?t=23096

**Comment by generalloki** [comment]
I beleive we could add style enum property into the ComboBox control with at least two predefined styles

```csharp
public enum ComboBoxKnownTheme
{
   Default,
   Button,
}
```

**Comment by Shadowblitz16** [comment]
> I beleive we could add style enum property into the ComboBox control with at least two predefined styles
> 
> ```cs
> public enum ComboBoxKnownTheme
> {
>    Default,
>    Button,
> }
> ```

ComboBox? Wouldn't RadioBox be better?

**Comment by generalloki** [comment]
Sure not ComboBox but RadioBox, my mistake. Thanks

**Comment by Shadowblitz16** [comment]
> Sure not ComboBox but RadioBox, my mistake. Thanks

It's ok I was just confirming

**Comment by Shadowblitz16** [comment]
@generalloki Is there a way to do this yet?
I need togglable buttons.

Maybe you could do what godot does and allow us to pass a button group.
Basicly this would just make every button in the button group act like radio buttons where only one button in the button group can be pressed at a time.. 

I need it for custom tabs.

**Comment by Shadowblitz16** [comment]
@generalloki I still need this, can we get something like a RadioButton and a CheckButton?

RadioButton would be a RadioBox that looks like a Button,
a CheckButton would be a CheckBox that looks like a Button

**Comment by generalloki** [comment]
I made public CardPanelHeader control, maybe you will find it useful for some of the cases. 

![image](https://github.com/user-attachments/assets/7b166e5f-99f5-4edc-969c-8ab0476e5a12)

```cs
           control.Add("Item 1");
           control.Add("Item 2");
           control.Add("Item 3");

           var item0 = control.Tabs[0];
           var item1 = control.Tabs[1];
           var item2 = control.Tabs[2];

           var size = 32;

           item0.Image = MessageBoxSvg.GetAsBitmap(MessageBoxIcon.Error, size, control);
           item1.Image = MessageBoxSvg.GetAsBitmap(MessageBoxIcon.Information, size, control);
           item2.Image = MessageBoxSvg.GetAsBitmap(MessageBoxIcon.Warning, size, control);

           control.TabClick += (s, e) =>
           {
               App.LogReplace(
                   $"CarPanelHeader.TabClick: {control.SelectedTabIndex}",
                   "CarPanelHeader.TabClick");
           };
```

It is possible to setup it like this:

![image](https://github.com/user-attachments/assets/0ed2a633-97ff-43bf-923f-939a4192a37a)

I will think what we can do about CheckBox and RadioButton custom painting theme

**Comment by generalloki** [comment]
Also in SpeedButton there are Sticky and StickyToggleOnClick properties. They allow to implement behavior like in CheckBox.

![image](https://github.com/user-attachments/assets/9c662d19-2700-4ab4-a7ce-84a580d7d042)

```cs
  c.Sticky = true; // Similar to IsChecked in CheckBox.
  c.StickyToggleOnClick = true;
  c.UseThemeForSticky = SpeedButton.KnownTheme.CheckBorder;
  c.UseTheme = SpeedButton.KnownTheme.StaticBorder;

 c.StickyChanged += (s, e) =>
 {
     App.LogReplace(
         $"SpeedButton.Sticky changed to [{c.Sticky}]",
         $"SpeedButton.Sticky");
 };
```
This require latest version from Master

**Comment by Shadowblitz16** [comment]
I am looking for a toggle and radio button that looks like a normal button, not customly drawn

**Comment by generalloki** [comment]
I believe we can add the option to draw standard push button background there, so it will look like Button.

**Comment by generalloki** [comment]
I have added three new themes to SpeedButton: ```PushButton, PushButtonHovered, PushButtonPressed```.
If SpeedButton is configured like this: 

```cs
                    c.Sticky = true; // Similar to IsChecked in CheckBox.
                    c.StickyToggleOnClick = true;
                    c.UseThemeForSticky = SpeedButton.KnownTheme.PushButtonHovered;
                    c.UseTheme = SpeedButton.KnownTheme.PushButton;
```

it will look like Button and behave like ```CheckBox```. Also if additionally setup SpeedButtons using ```StickySpreadMode```

```cs
            item1.StickySpreadMode = CheckedSpreadMode.SingleInGroup;
            item1.GroupIndex = 53; // some custom group index

            item2.StickySpreadMode = CheckedSpreadMode.SingleInGroup;
            item2.GroupIndex = 53;
```

they will behave like ```RadioButton```. There is also ```CheckedSpreadMode.SingleSibling``` which is the similar to
CheckedSpreadMode.SingleInGroup, but doesn't require GroupIndex and works within Parent.

**Comment by Shadowblitz16** [comment]
ok ty.

**Comment by Shadowblitz16** [comment]
@generalloki this might be good as separate controls.

Checkbox -> Checkbutton
Radiobox -> Radiobutton 
Combobox -> Combobutton
Listbox -> Listbutton
Spinbox -> Spinbutton

Checkbutton and Radiobutton would be a Checkbox and Radiobox in the style of a button.

Combobutton, Listbutton, and Spinbutton would in the style of a button but when you click on it it goes to the next item, wrapping around if it reaches the end.

We could then get a ComboCarasole, ListCarasole, and SpinCarasole which are like the buttons equivalents but have a left and right button on the left and right side of the control similar to AAA games option menus.

**Reply by generalloki** [reply]
We are planning to have StdCheckBox and StdRadioButton which will not use native controls and will be implemented inside the library. After they are implemented we may consider adding there something like ThemeKind so it will be possible to specify how their look (Default, AsButton, etc.). Currently, I can only suggest to use SpeedButton with properties configuration as I mentioned in my above message. 

As to other controls, I think each control that you suggest might be in the separate issue with at least 2-3 sentence description on how it should look and behave. And possibly with some images.
